lib/Solver/Z3Builder.cpp

```diff
@@ -400,16 +401,18 @@ Z3ASTHandle Z3Builder::getInitialArray(const Array *root) {
     array_expr = buildArray(unique_name.c_str(), root->getDomain(),
                             root->getRange());
 
-    if (root->isConstantArray()) {
-      // FIXME: Flush the concrete values into Z3. Ideally we would do this
-      // using assertions, which might be faster, but we need to fix the caching
-      // to work correctly in that case.
-      for (unsigned i = 0, e = root->size; i != e; ++i) {
-        Z3ASTHandle prev = array_expr;
-        array_expr = writeExpr(
-            prev, construct(ConstantExpr::alloc(i, root->getDomain()), 0),
-            construct(root->constantValues[i], 0));
+    if (root->isConstantArray()
+        && constant_array_assertions.count(root) == 0) {
+     std::vector<Z3ASTHandle> array_assertions;
+     for (unsigned i = 0, e = root->size; i != e; ++i) {
+        //construct(= (select i root) root->value[i]) to be asserted in Z3Solver.cpp
+        array_assertions.emplace_back(
+          Z3_mk_eq(ctx,
+                   Z3_mk_select(ctx, array_expr, bvConst32(root->getDomain(), i)),
+                   construct(root->constantValues[i],0)
+          ), ctx);
       }
+      constant_array_assertions[root] = array_assertions;
```

Comment:
You should probably use `std::move()` (or add the vector to the map first and then get a reference to it) to avoid copying `array_assertions` and then destroying the original.

Reply:
I thought compiler can figure this out, but I'll add it.